Origin: https://github.com/php/php-src/commit/7ac9700d0d46a7b96e72e330efd295a3332124b0
Advisory: GHSA-ch8v-r6jh-4vvr
Imported-by: mammoth-lts 7.2 backport (patches/7.2/9007-GHSA-ch8v-r6jh-4vvr-sanitize-encoded-0xff.patch), applies as-is to 7.1

From 6b997741fb1fa7674d55127ac84e7aea1e45b7b1 Mon Sep 17 00:00:00 2001
From: Ilia Alshanetsky <ilia@ilia.ws>
Date: Wed, 29 Jul 2026 11:21:50 -0400
Subject: [PATCH] Fix GHSA-ch8v-r6jh-4vvr: encode 0xFF in
 FILTER_SANITIZE_ENCODED

php_filter_encode_url() initialized its 256-byte "must encode" table with
memset(tmp, 1, sizeof(tmp) - 1), leaving tmp[255] uninitialized. Whether
0xFF got percent-encoded then depended on stack garbage; valgrind reports
the read as a conditional jump on an uninitialised value. Initialize the
whole table.

Backport of 56149186d81469fd7fb2f8444359ee81dd4a70f4 from PHP-8.4.

Backport of php-src 7ac9700d0d46a7b96e72e330efd295a3332124b0 (PHP-8.2) to PHP 7.2
---
 ext/filter/sanitizing_filters.c                    |  2 +-
 ext/filter/tests/filter_sanitize_encoded_0xff.phpt | 12 ++++++++++++
 2 files changed, 13 insertions(+), 1 deletion(-)
 create mode 100644 ext/filter/tests/filter_sanitize_encoded_0xff.phpt

diff --git a/ext/filter/sanitizing_filters.c b/ext/filter/sanitizing_filters.c
index 9b0a3aaa..c2e06651 100644
--- a/ext/filter/sanitizing_filters.c
+++ b/ext/filter/sanitizing_filters.c
@@ -71,7 +71,7 @@ static void php_filter_encode_url(zval *value, const unsigned char* chars, const
 	unsigned char *e = s + char_len;
 	zend_string *str;
 
-	memset(tmp, 1, sizeof(tmp)-1);
+	memset(tmp, 1, sizeof(tmp));
 
 	while (s < e) {
 		tmp[*s++] = '\0';
diff --git a/ext/filter/tests/filter_sanitize_encoded_0xff.phpt b/ext/filter/tests/filter_sanitize_encoded_0xff.phpt
new file mode 100644
index 00000000..1ee61aba
--- /dev/null
+++ b/ext/filter/tests/filter_sanitize_encoded_0xff.phpt
@@ -0,0 +1,12 @@
+--TEST--
+FILTER_SANITIZE_ENCODED percent-encodes 0xFF
+--EXTENSIONS--
+filter
+--FILE--
+<?php
+var_dump(filter_var("\xFF", FILTER_SANITIZE_ENCODED));
+var_dump(filter_var("\xFE\xFF\x00A", FILTER_SANITIZE_ENCODED));
+?>
+--EXPECT--
+string(3) "%FF"
+string(10) "%FE%FF%00A"
-- 
2.52.0

